File: 백준/Silver/10816. 숫자 카드 2/숫자 카드 2.py
# 10
# 6 3 2 10 10 10 -10 -10 7 3
# 8
# 10 9 -5 2 3 4 5 -10
import sys
from bisect import bisect_left, bisect_right
input = sys.stdin.readline

# 1) 입력 받기
N = int(input())
lst_N = sorted(list(map(int, input().split())))
M = int(input())
lst_M = list(map(int, input().split()))

answer = []

# 2) binary search
def count_by_range(left_value, right_value):
    right_index = bisect_right(lst_N, right_value)
    left_index = bisect_left(lst_N, left_value)
    return right_index - left_index

# 3) 반복
for i in range(M):
    num = count_by_range(lst_M[i], lst_M[i])
    answer.append(num)

print(*answer)

File: 백준/Silver/10816. 숫자 카드 2/숫자 카드 2.js
const INPUT = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

function solution() {
  const N = Number(INPUT[0]);
  const lst_N = INPUT[1]
    .split(" ")
    .map(Number)
    .sort((a, b) => a - b);
  const M = Number(INPUT[2]);
  const lst_M = INPUT[3].split(" ").map(Number);

  let answer = "";

  const lowerBound = (target, start, end) => {
    while (start < end) {
      let mid = parseInt((start + end) / 2);
      if (lst_N[mid] >= target) end = mid;
      else start = mid + 1;
    }
    return end;
  };

  const upperBound = (target, start, end) => {
    while (start < end) {
      let mid = parseInt((start + end) / 2);
      if (lst_N[mid] > target) end = mid;
      else start = mid + 1;
    }
    return end;
  };

  const countByRange = (leftValue, rightValue) => {
    let rightIndex = upperBound(rightValue, 0, lst_N.length);
    let leftIndex = lowerBound(leftValue, 0, lst_N.length);
    return rightIndex - leftIndex;
  };

  for (let i = 0; i < M; i++) {
    const num = countByRange(lst_M[i], lst_M[i]);
    answer += `${num} `;
  }

  console.log(answer);
}

solution();